go/store/nbs/generational_chunk_store.go

```diff
@@ -0,0 +1,302 @@
+		return nil, err
+	}
+
+	if len(notInOldGen) == 0 {
+		return notInOldGen, nil
+	}
+
+	return gcs.newGen.HasMany(ctx, notInOldGen)
+}
+
+// Put caches c in the ChunkSource. Upon return, c must be visible to
+// subsequent Get and Has calls, but must not be persistent until a call
+// to Flush(). Put may be called concurrently with other calls to Put(),
+// Get(), GetMany(), Has() and HasMany().
+func (gcs *GenerationalNBS) Put(ctx context.Context, c chunks.Chunk) error {
+	return gcs.newGen.Put(ctx, c)
+}
+
+// Returns the NomsVersion with which this ChunkSource is compatible.
+func (gcs *GenerationalNBS) Version() string {
+	return gcs.newGen.Version()
+}
+
+// Rebase brings this ChunkStore into sync with the persistent storage's
+// current root.
+func (gcs *GenerationalNBS) Rebase(ctx context.Context) error {
+	oErr := gcs.oldGen.Rebase(ctx)
+	nErr := gcs.newGen.Rebase(ctx)
+
+	if oErr != nil {
+		return oErr
+	}
+
+	return nErr
+}
+
+// Root returns the root of the database as of the time the ChunkStore
+// was opened or the most recent call to Rebase.
+func (gcs *GenerationalNBS) Root(ctx context.Context) (hash.Hash, error) {
+	return gcs.newGen.Root(ctx)
+}
+
+// Commit atomically attempts to persist all novel Chunks and update the
+// persisted root hash from last to current (or keeps it the same).
+// If last doesn't match the root in persistent storage, returns false.
+func (gcs *GenerationalNBS) Commit(ctx context.Context, current, last hash.Hash) (bool, error) {
+	return gcs.newGen.Commit(ctx, current, last)
+}
+
+// Stats may return some kind of struct that reports statistics about the
+// ChunkStore instance. The type is implementation-dependent, and impls
+// may return nil
+func (gcs *GenerationalNBS) Stats() interface{} {
+	return nil
+}
+
+// StatsSummary may return a string containing summarized statistics for
+// this ChunkStore. It must return "Unsupported" if this operation is not
+// supported.
+func (gcs *GenerationalNBS) StatsSummary() string {
+	return ""
+}
+
+// Close tears down any resources in use by the implementation. After // Close(), the ChunkStore may not be used again. It is NOT SAFE to call
+// Close() concurrently with any other ChunkStore method; behavior is
+// undefined and probably crashy.
+func (gcs *GenerationalNBS) Close() error {
+	oErr := gcs.oldGen.Close()
+	nErr := gcs.newGen.Close()
+
+	if oErr != nil {
+		return oErr
+	}
+
+	return nErr
+}
+
+func (gcs *GenerationalNBS) copyToOldGen(ctx context.Context, hashes hash.HashSet) error {
+	notInOldGen, err := gcs.oldGen.HasMany(ctx, hashes)
+
+	if err != nil {
+		return err
+	}
+
+	var putErr error
+	err = gcs.newGen.GetMany(ctx, notInOldGen, func(chunk *chunks.Chunk) {
+		if putErr == nil {
+			putErr = gcs.oldGen.Put(ctx, *chunk)
+		}
+	})
+
+	if putErr != nil {
+		return putErr
+	}
+
+	return err
+}
+
+func (gcs *GenerationalNBS) MarkAndSweepChunks(ctx context.Context, last hash.Hash, keepChunks <-chan []hash.Hash, dest chunks.ChunkStore) error {
```

Comment:
this type doesn't need to satisfy the `GarbageCollector` interface, we can delete this

Reply:
good call